domaci_maze.py

```diff
@@ -90,495 +77,523 @@ def has_value(self) -> bool:
+    values = init_q_values(env) if q_function else init_v_values(env)
+    for iteration in range(maxit):
+        # print(f"\nIteration: {iteration + 1}")
+        # print("Old values: ", values)
+        values_copy = copy(values)
+        new_values = async_update_all_values(env, values, gamma, q_function)
+        # print("New values: ", new_values)
+        err = max([abs(new_values[s] - values_copy[s]) for s in values])
         if err < eps:
-            return nv, k
-        v = nv
-    return v, k
-
+            # print("Final error is: ", err)
+            return new_values, iteration + 1
+        # print("Error is: ", err)
+        values = new_values
+    return values, iteration + 1
+
+
+# should return the smallest number action if there are more actions that have the same value
+# if values of both actions 1 and 2 are -1.0 , it returns 1
+def best_action_min_arg(actions_probs):
+    max_probability = max(prob for _, prob in actions_probs)
+    max_probability_elements = [(action, prob) for action, prob in actions_probs if prob == max_probability]
+
+    min_action = min(action for action, _ in max_probability_elements)
+    min_action_element = [(action, prob) for action, prob in max_probability_elements if action == min_action][0][0]
+
+    return min_action_element
+
+
+def greedy_action(env, current_node, values, gamma, q_function=False):
+    if q_function:
+        action_values = []
+        for action in ACTIONS:
+            for next_node, prob in env.get_action_probabilities(current_node, action):
+                possible_values_one_node = []
+                for next_action in ACTIONS:
+                    possible_values_one_node.append(values[next_node.get_position(), next_action])
+                action_values.append((action, prob * (next_node.get_reward() + gamma * max(possible_values_one_node))))
+        return best_action_min_arg(action_values) if action_values else None
+        # return max(action_values, key=lambda x: x[1])[0] if action_values else None
+    else:
+        action_values = []
+        for action in ACTIONS:
+            temp = 0
+            for next_node, prob in env.get_action_probabilities(current_node, action):
+                temp += prob * (next_node.get_reward() + gamma * values[next_node.get_position()])
+            action_values.append((action, temp))
+        return best_action_min_arg(action_values) if max(action_values) != 0 else None
+        # return max(action_values, key=lambda x: x[1])[0] if max(action_values) != 0 else None
 
-fin_v, k = value_iteration(env, 1.0, 0.01)
-draw_values(env, fin_v)
-plt.title(f"Converged after {k} iterations")
 
+def generate_optimal_policy(env, values, gamma, q_function=False):
+    return {
+        node.get_position(): greedy_action(env, node, values, gamma, q_function)
+        for node in env.get_graph()
+        if not (node.is_terminal() or not node.is_steppable())
+    }

```

Comment:
Sta se desi u else? Sta onda bude u dictionary? Nista?

Reply:
if not (node.is_terminal() or not node.is_steppable()) vraca True za sve sem WallNode i Terminal node za koje nam ni ne treba politika jer nema nigde iz njih da se ide. Znaci pravi optimalnu politiku za sve sem 2 gore navedena.